cmake/FindCUDACompilerNVCC.cmake

```diff
@@ -0,0 +1,48 @@
+# Find the NVCC compiler, include its libraries and declare a custom
+# `add_library()` wrapper function named `add_gpu_library()`.
+
+set(CMAKE_CUDA_COMPILER ${CUDA_NVCC_EXECUTABLE})
+set(CUDA 1)
+
+execute_process(COMMAND ${CMAKE_CUDA_COMPILER} --version
+                OUTPUT_VARIABLE NVCC_VERSION_STRING)
+
+string(REGEX
+       REPLACE "^.*Cuda compilation tools, release [0-9\.]+, V([0-9\.]+).*\$"
+               "\\1"
+               CMAKE_CUDA_COMPILER_VERSION
+               "${NVCC_VERSION_STRING}")
+
+if(NOT CUDA_NVCC_EXECUTABLE STREQUAL "${CUDA_TOOLKIT_ROOT_DIR}/bin/nvcc")
+  get_filename_component(NVCC_EXECUTABLE_DIRNAME "${CUDA_NVCC_EXECUTABLE}" DIRECTORY)
+  get_filename_component(NVCC_EXECUTABLE_DIRNAME "${NVCC_EXECUTABLE_DIRNAME}" DIRECTORY)
+  message(WARNING "Your nvcc (${CUDA_NVCC_EXECUTABLE}) does not appear to match your CUDA libraries (in ${CUDA_TOOLKIT_ROOT_DIR}). While ESPResSo will still compile, you might get unexpected crashes. Please point CUDA_TOOLKIT_ROOT_DIR to your CUDA toolkit path, e.g. by adding -DCUDA_TOOLKIT_ROOT_DIR='${NVCC_EXECUTABLE_DIRNAME}' to your cmake command.")
+endif()
+
+set(CUDA_LINK_LIBRARIES_KEYWORD PUBLIC)
+
+SET(CUDA_PROPAGATE_HOST_FLAGS OFF)
+
+set(CUDA_NVCC_FLAGS_DEBUG "${CUDA_NVCC_FLAGS_DEBUG} -g")
+set(CUDA_NVCC_FLAGS_RELEASE "${CUDA_NVCC_FLAGS_RELEASE} -O3 -DNDEBUG")
+set(CUDA_NVCC_FLAGS_MINSIZEREL "${CUDA_NVCC_FLAGS_MINSIZEREL} -Os -DNDEBUG")
```

Comment:
several issues here:
* `-Os` is not a valid optimization level: `nvcc fatal   : 's': expected a number`
* there seems to be a difference between `-O2`, `-Xptxas -O2`, `-Xcompiler -O2` ([StackOverflow](https://stackoverflow.com/a/43845984))

Reply:
progress:
* CUDA compiler flags are actually quite complicated. Also, device and host compiler optimization flags don't mix well.
* We don't need to touch the host and device flags in `build_cmake.sh`. The logic is wrong anyway, and it is partially overridden by CMake in Debug builds.

I'm currently removing a lot of useless code in `build_cmake.sh`, but getting the CMake logic right will take me a few more days, I'm afraid. After that, we'll be able to experiment with CMake pseudo-interfaces for GPU targets.

Reply:
I wasted two man hours investigating a Clang compiler error, because when we do
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/src/core/CMakeLists.txt#L147
we forward CUDA compiler flags to regular .cpp files:
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/CMakeLists.txt#L165
even though a few lines later there's a mechanism to set those flags to .cu files only:
https://github.com/espressomd/espresso/blob/d67ab4aefd263997cbec82015052a19f48b48385/CMakeLists.txt#L172

Reply:
I wasted one man hour because neither `CMakeLists.txt` nor `build_cmake.sh` throw an error when using the incompatible combination of CMake flags `-DCUDA_NVCC_EXECUTABLE=$(which clang++) -DWITH_COVERAGE=ON`.

Reply:
My last PR introduced a regression: the boost version depends on the CUDA version, so CUDA must be loaded before boost...

Reply:
Why do we override C++ flags only for release builds and not all other types of build???
https://github.com/espressomd/espresso/blob/f291cf448f30b28488830e65556436659870bbf7/src/core/CMakeLists.txt#L78-L79

Reply:
Why `-01` specifically? And why only for release builds?
https://github.com/espressomd/espresso/blob/f291cf448f30b28488830e65556436659870bbf7/CMakeLists.txt#L347-L348

Reply:
FindCUDA didn‘t really have a concept of debug vs. release, so we manually set these flags to match the host compiler and get debuggable or optimized CUDA code as requested.

Reply:
> FindCUDA didn‘t really have a concept of debug vs. release

doesn't it since 3.0? [`CUDA_NVCC_FLAGS_<CONFIG>`](https://cmake.org/cmake/help/v3.0/module/FindCUDA.html?highlight=CUDA_NVCC_FLAGS_%3CCONFIG%3E)

HIP also does: [`HIP_<COMPILER>_FLAGS_<CONFIG>`](https://github.com/ROCm-Developer-Tools/HIP/blob/821c60a3d98f5f4d6cda9440c1452f73edb93b5a/cmake/FindHIP.cmake#L15-L19)

Reply:
I believe that didn't set the flags for both host and device, but that may no longer be true. The `-g -O1` for `WITH_ASAN` is a compromise between good sanitizability and reasonable performance.